hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,384 @@
+
+    private final HoodieTableMetadataWriter metadataWriter;
+    private final List<HoodieInstant> instantsToIndex;
+    private final Set<String> metadataCompletedInstants;
+    private final HoodieTableMetaClient metaClient;
+    private final HoodieTableMetaClient metadataMetaClient;
+
+    IndexingCatchupTask(HoodieTableMetadataWriter metadataWriter,
+                        List<HoodieInstant> instantsToIndex,
+                        Set<String> metadataCompletedInstants,
+                        HoodieTableMetaClient metaClient,
+                        HoodieTableMetaClient metadataMetaClient) {
+      this.metadataWriter = metadataWriter;
+      this.instantsToIndex = instantsToIndex;
+      this.metadataCompletedInstants = metadataCompletedInstants;
+      this.metaClient = metaClient;
+      this.metadataMetaClient = metadataMetaClient;
+    }
+
+    @Override
+    public void run() {
+      for (HoodieInstant instant : instantsToIndex) {
+        // metadata index already updated for this instant
+        if (!metadataCompletedInstants.isEmpty() && metadataCompletedInstants.contains(instant.getTimestamp())) {
+          currentIndexedInstant = instant.getTimestamp();
+          continue;
+        }
+        while (!instant.isCompleted()) {
+          try {
+            LOG.warn("instant not completed, reloading timeline " + instant);
+            // reload timeline and fetch instant details again wait until timeout
+            String instantTime = instant.getTimestamp();
+            Option<HoodieInstant> currentInstant = metaClient.reloadActiveTimeline()
+                .filterCompletedInstants().filter(i -> i.getTimestamp().equals(instantTime)).firstInstant();
+            instant = currentInstant.orElse(instant);
+            // so that timeline is not reloaded very frequently
+            Thread.sleep(TIMELINE_RELOAD_INTERVAL_MILLIS);
+          } catch (InterruptedException e) {
+            throw new HoodieIndexException(String.format("Thread interrupted while running indexing check for instant: %s", instant), e);
+          }
+        }
+        // if instant completed, ensure that there was metadata commit, else update metadata for this completed instant
+        if (COMPLETED.equals(instant.getState())) {
+          String instantTime = instant.getTimestamp();
+          Option<HoodieInstant> metadataInstant = metadataMetaClient.reloadActiveTimeline()
+              .filterCompletedInstants().filter(i -> i.getTimestamp().equals(instantTime)).firstInstant();
+          if (metadataInstant.isPresent()) {
+            currentIndexedInstant = instantTime;
+            continue;
+          }
+          try {
+            // we need take a lock here as inflight writer could also try to update the timeline
+            txnManager.beginTransaction(Option.of(instant), Option.empty());
+            LOG.info("Updating metadata table for instant: " + instant);
+            switch (instant.getAction()) {
```

Comment:
is n't there a top level method in metadata writer to handle different instant types ? We can reuse that or move this code there

Reply:
no, will take this refactoring along with a followup task.. it needs a little bit more than extracting to a method.